src/packed-pack.ts

```diff
@@ -283,19 +280,101 @@ export class PackContent {
+            objectsByType[btype].unpackedSize += unpackedSize;
+
+            if (isOffsetDeltaPrelude(objectPrelude)) {
+                objectsByType[btype].unpackedRestoredSize += objectPrelude.unpackedRestoredSize;
+            }
         }
 
         return objectsByType.filter((stat) => stat.count > 0);
     }
+
+    private read(buffer: Buffer, offset: number) {
+        return this.fh.read(buffer, 0, buffer.byteLength, offset);
+    }
+
+    private async readObjectPreludeFromFile(
+        offset: number
+    ): Promise<[BufferCursor, ObjectPrelude]> {
+        const header =
+            reuseBufferCount > 0 ? buffers[--reuseBufferCount] : Buffer.allocUnsafe(4096);
+
+        await this.read(header, offset);
+        const reader = new BufferCursor(header);
+
+        return [reader, await this.readObjectPrelude(offset, reader)];
+    }
+
+    private async readObjectPrelude(
+        offsetBase: number,
+        reader: BufferCursor
+    ): Promise<ObjectPrelude> {
+        // object start relative to current buffer
+        const relOffset = reader.offset;
+
+        // https://git-scm.com/docs/pack-format#_size_encoding
+        // n-byte type and length (3-bit type, (n-1)*7+4-bit length) compressed data
+        const int = readVarIntLE(reader);
+        const btype = ((int >> 4) & 0b0111) as PackedType;
+        const length = ((int >> 7) << 4) | (int & 0b1111);
+
+        if (btype === OBJ_INVALID || btype === OBJ_RESERVED) {
+            throw new Error(`Unrecognized type: 0b${btype.toString(2)}`);
+        }
+
+        const result = {
+            btype,
+            unpackedSize: length
+        } as ObjectPrelude;
+
+        if (btype !== OBJ_OFS_DELTA && btype !== OBJ_REF_DELTA) {
+            return result as BaseObjectPrelude;
+        }
+
+        // Handle deltified objects
+        // Both ofs-delta and ref-delta store the "delta" to be applied to another object (called base object)
+        // to reconstruct the object. The difference between them is, ref-delta directly encodes base object name.
+        // If the base object is in the same pack, ofs-delta encodes the offset of the base object in the pack instead.
+        if (isOffsetDeltaPrelude(result)) {
+            result.deltaRef = relOffset + offsetBase - readEncodedOffset(reader);
+        } else if (isRefDeltaPrelude(result)) {
+            result.deltaRef = Buffer.from(reader.slice(20));
+        }
+
+        if (!isOffsetDeltaPrelude(result)) {
+            return result;
+        }
+
+        const chunkSize = 512;
+        const objectBuffer = Buffer.allocUnsafe(chunkSize);
+
+        if (reader.bytesLeft >= chunkSize) {
+            const offsetBefore = reader.offset;
+            reader.copyTo(objectBuffer);
+            // restore original offset to not disturb reader usages outside the function
+            reader.offset = offsetBefore;
+        } else {
+            await this.read(objectBuffer, offsetBase + reader.offset);
```

Comment:
`Buffer.allocUnsafe(chunkSize)` should be performed only in this case, otherwise reader's buffer is reused via a slice.